launch_ros/launch_ros/actions/node.py

```diff
@@ -434,6 +445,55 @@ def _perform_substitutions(self, context: LaunchContext) -> None:
                 cmd_extension.extend(['-r', f'{src}:={dst}'])
             self.cmd.extend([normalize_to_list_of_substitutions(x) for x in cmd_extension])
 
+    def _secure_self(
+        self, context: LaunchContext, ros_specific_arguments: Dict[str, Union[str, List[str]]]
+    ):
+        """Enable encryption, creating a key for the node if necessary."""
+        nodl_node = nodl.get_node_by_executable(
+            package_name=self.__package, executable_name=self.__node_executable
+        )
+
+        self.__enclave = self.node_name.replace(
+            Node.UNSPECIFIED_NODE_NAME, nodl_node.name
+        ).replace(Node.UNSPECIFIED_NODE_NAMESPACE, '')
+
+        self.__keystore_path = os.environ.get('ROS_SECURITY_KEYSTORE', None)
+
+        if not(sros2.api._keystore.is_valid_keystore(self.__keystore_path)):
+            if not context.launch_configurations.get('__no_create_keystore', None):
+                # If keystore path is blank, create a transient keystore
+                if not self.__keystore_path:
+                    self.__keystore_path = self._create_transient_keystore(context)
+
+                # If keystore is not initialized, create a keystore
+                if not pathlib.Path(self.__keystore_path).is_dir():
+                    pathlib.Path(self.__keystore_path).mkdir()
+                sros2.api._keystore.create_keystore(self.__keystore_path)
+            else:
+                if not self.__keystore_path:
+                    raise RuntimeError('--no-create-keystore enabled, but no keystore specified!')
+                else:
+                    raise RuntimeError(('--no-create-keystore was specified and '
+                                        f'ROS_SECURITY_KEYSTORE="{self.__keystore_path}" '
+                                        'is not a valid keystore'))
+
+        if not sros2.api._key.create_key(
+            keystore_path=self.__keystore_path, identity=self.__enclave
+        ):
+            raise RuntimeError(
```

Comment:
I really don't like generic exceptions, is this the established pattern in launch_ros, or do they have a set of defined exceptions?

Regardless, would it be more helpful if this error actually said what was wrong? Mention something about being unable to create a key?

Reply:
This is the established pattern, they only use RuntimeErrors

Reply:
Okay. Ignoring that gripe then, how about making the message more helpful?

Reply:
No longer relevant